lib/primitives/mtx.js

```diff
@@ -868,6 +868,51 @@ class MTX extends TX {
     return null;
   }
 
+  /**
+   * Sign custom P2WSH input.
+   * @param {Number} index - Index of input being signed.
+   * @param {KeyRing[]} rings - with private key.
+   * Sorted by desired signature order.
+   * @param {Script} redeemScript - its hash is in spending output.
+   * @param {Buffer|Number[]} witnessScript - signatures template.
+   * witnessScript is the SegWit equivalent of scriptSig.
+   * Put an empty buffer as placeholder for each signature.
+   * @param {SighashType} type
+   * @param {Coin|Output|Number} coin or output or value.
+   * @returns {Boolean} Whether the input was able to be signed.
+   */
+
+  signCustomInput(index, rings, redeemScript, witnessScript, type, coin) {
+    function getValue(tx, coin, input) {
+      if (!coin)
+        coin = tx.view.getOutputFor(input);
+
+      assert(coin, 'Cannot get value of spending output');
+
+      if (Number.isInteger(coin))
+        return coin;
+
+      return coin.value;
+    }
+
+    const version = 1;
+    const input = this.inputs[index];
+    const value = getValue(this, coin, input);
+
+    const sigs = [];
+    for (const ring of rings) {
+      sigs.push(this.signature(index, redeemScript, value,
```

Comment:
The rings passed into the function will have to be in the correct order to match the redeem script. (Swapping `ring1` and `ring2` in your tests will fail them). I wonder if we should find the right place for each key/sig on behalf of the caller. Similar method is employed in `mtx.signVector()`:

https://github.com/bcoin-org/bcoin/blob/67d444b5ec45e45147bc780573494334c1ef3019/lib/primitives/mtx.js#L811-L821

Reply:
Also note that `mtx.signVector()` will throw if a key passed in does not sign anything in the tx. If you add a third KeyRing `ring3` in your test and pass it to `mtx.signCustomInput()` there will be no warning to the user, just a failed `mtx.verify()`

Reply:
I'm afraid that to achieve this we'll either use heuristics, or we'll have to cook up some algorithm which will have to actually run the script and try out different signatures in different spots, but maybe running the script is too expensive for what it gives us.

For example, if our heuristic for checking if there are redundant rings is "the public key of all rings must appear in the script", then it would reject a P2PKH. We would also have a hard time recognizing that the correct order in this script
```
OP_IF
  <pubkey1>
  OP_CHECKSIGVERIFY
OP_ELSE
  <pubkey2>
  OP_CHECKSIGVERIFY
OP_ENDIF
<pubkey3>
OP_CHECKSIG
```
is either '`sig3` `sig2`' or '`sig3` `sig1`' without actually running the script. Then a user that thinks "bcoin says it's smart, it will figure it out itself" would be surprised to find out that when she gives `[ring1, ring2, ring3]` as inputs, the function would (maybe) naively sign with the first two matching rings to produce the scriptSig '`0`|`1` `sig2` `sig1`', which is always wrong.

What we could do instead is run `this.checkInput()` after signing in order to loudly fail if signing didn't work. I've implemented this for now, but unfortunately it has witness logic hardcoded.

Let me know if you think it's worth it to actually run the script.